tests/cpp/test_multidevice_transformer.cpp

```diff
@@ -1095,17 +1062,16 @@ TEST_P(DistributedTransformerTest, Backward) {
   TensorView* ln0_b = makeContigTensor(1);
   TensorView* ln0_mean = makeConcreteTensor({B * S, 1});
   TensorView* ln0_rstd = makeConcreteTensor({B * S, 1});
+  TensorView* mha_linear0 = makeContigTensor(3, dtype);
   TensorView* mha_linear1 = makeContigTensor(2);
```

Comment:
Should this be dtype as well?

Reply:
Ideally, yes mha_linear1 would also be dtype. But, we compute mha_linear1 as: local matmul (bfloat16) + allreduce (bfloat16) + sum (**float**) because nvfuser doesn't support bfloat16 reduction. 

I will open an issue about supporting bfloat16 reduction, since this is a pattern we will see. We can either (1) keep mha_linear1 as float to avoid the cast from bfloat16->float or (2) change the dtype and add the casts.

Reply:
nvFuser doesn't support bfloat16 reduction **in codegen**. But I believe it supports bfloat16 reduction when it's **lowered to communication**. 

For example, https://github.com/NVIDIA/Fuser/blob/31d3a629bb954cec421404fea79a8aba362fb961/tests/python/test_multidevice.py#L423 has been using bfloat16 reduction just fine. I further confirmed this by running

```bash
$ git checkout wjy/shard
$ _bn && mpirun -np 2 -x NVFUSER_DUMP=fusion_ir_preseg pytest tests/python/test_multidevice.py -k 'backward' --only-mpi -s > /tmp/stdout
$ grep -B 1 -a -e 'reduction.*bfloat' /tmp/stdout
T65_l___bfloat[ rS212{2}, iS213{2048}, iS214{12288} ] (DeviceMesh{0 1})
   = reduction( T64_l___bfloat[ ideviceIdx.x208{2}, iS209{2048}, iS210{12288}, rS211{24576} ] (DeviceMesh{0 1}), op = add, initial value = __bfloat(0), allreduce = false )
--
T155_l___bfloat[ rS536{2}, iS537{2048}, iS538{12288} ] (DeviceMesh{0 1})
   = reduction( T154_l___bfloat[ ideviceIdx.x532{2}, iS533{2048}, iS534{12288}, rS535{18432} ] (DeviceMesh{0 1}), op = add, initial value = __bfloat(0), allreduce = false )
```

(I'm totally fine with pursing that in a separate PR)

Reply:
My bad, it's not the reduction that's the issue (this is in bfloat16 as you mentioned) it's the bias add.